promql/engine.go

```diff
@@ -615,88 +575,168 @@ func (ev *evaluator) recover(errp *error) {
-	if !ok {
-		ev.errorf("expected string but got %s", documentedType(val.Type()))
+	h := l.Hash()
+	ret, ok := enh.dmn[h]
+	if ok {
+		return ret
 	}
-	return sv
+	ret = dropMetricName(l)
+	enh.dmn[h] = ret
+	return ret
 }
 
-// evalOneOf evaluates e and errors unless the result is of one of the given types.
-func (ev *evaluator) evalOneOf(e Expr, t1, t2 ValueType) Value {
-	val := ev.eval(e)
-	if val.Type() != t1 && val.Type() != t2 {
-		ev.errorf("expected %s or %s but got %s", documentedType(t1), documentedType(t2), documentedType(val.Type()))
+// signatureFunc is a cached version of signatureFunc.
+func (enh *evalNodeHelper) signatureFunc(on bool, names ...string) func(labels.Labels) uint64 {
+	if enh.sigf == nil {
+		enh.sigf = make(map[uint64]uint64, len(enh.out))
+	}
+	f := signatureFunc(on, names...)
+	return func(l labels.Labels) uint64 {
+		h := l.Hash()
+		ret, ok := enh.sigf[h]
+		if ok {
+			return ret
+		}
+		ret = f(l)
+		enh.sigf[h] = ret
+		return ret
+	}
+}
+
+// rangeEval evaluates the given expressions, and then for each step calls
+// the given function with the values computed for each expression at that
+// step.  The return value is the combination into time series of  of all the
+// function call results.
+func (ev *evaluator) rangeEval(f func([]Value, *evalNodeHelper) Vector, exprs ...Expr) Matrix {
+	numSteps := int((ev.endTimestamp-ev.timestamp)/ev.interval) + 1
+	matrixes := make([]Matrix, len(exprs))
+	origMatrixes := make([]Matrix, len(exprs))
+	for i, e := range exprs {
+		// Functions will take string arguments from the expressions, not the values.
+		if e != nil && e.Type() != ValueTypeString {
+			matrixes[i] = ev.eval(e).(Matrix)
+
+			// Keep a copy of the original point slices so that they
+			// can be returned to the pool.
+			origMatrixes[i] = make(Matrix, len(matrixes[i]))
+			copy(origMatrixes[i], matrixes[i])
+		}
 	}
-	return val
-}
 
-func (ev *evaluator) Eval(expr Expr) (v Value, err error) {
-	defer ev.recover(&err)
-	defer ev.close()
-	return ev.eval(expr), nil
+	vectors := make([]Vector, len(exprs)) // Input vectors for the function.
+	args := make([]Value, len(exprs))     // Argument to function.
+	// Create an output vector that is as big as the input matrix with
+	// the most time series.
+	biggestLen := 1
+	for i := range exprs {
+		vectors[i] = make(Vector, 0, len(matrixes[i]))
+		if len(matrixes[i]) > biggestLen {
+			biggestLen = len(matrixes[i])
+		}
+	}
+	enh := &evalNodeHelper{out: make(Vector, 0, biggestLen)}
+	seriess := make(map[uint64]Series, biggestLen) // Output series by series hash.
+	for ts := ev.timestamp; ts <= ev.endTimestamp; ts += ev.interval {
+		// Gather input vectors for this timestamp.
+		for i := range exprs {
+			vectors[i] = vectors[i][:0]
+			for si, series := range matrixes[i] {
+				for pos, point := range series.Points {
+					if point.T == ts {
+						vectors[i] = append(vectors[i], Sample{Metric: series.Metric, Point: point})
+					}
+					if point.T >= ts {
```

Comment:
We can simplify a little here, but still need the 2nd if to deal with a time series that only appears half way through efficiently.

You can't have just `series.Points` on the LHS, as that's only a local variable so doesn't actually reduce the size of the slice for the next round.

Reply:
> We can simplify a little here, but still need the 2nd if to deal with a time series that only appears half way through efficiently.

Good catch, but that's only because there's a `for` loop. Now that I look at it, it seems to me you could drop the for loop altogether: the point at timestamp `ts` is either there or it's missing, in which case `point.T` is already past `ts`.

> You can't have just `series.Points` on the LHS, as that's only a local variable so doesn't actually reduce the size of the slice for the next round.

I'm still struggling with' Go's value vs. reference thing (I think they should have kept the `*` for clarity), so still only wondering: is `series` a reference to or a copy of the series in the matrix? Because if it's the former using `series.Points` as the LHS should work just fine. Else I really need to re-read the Go documentation.

Reply:
Figured out the value vs. reference and you are of course right. Apologies for the confusion.

The `for` loop though, still appears to be unnecessary.

Reply:
We can remove the 2nd if and unconditionally break, the for handles the case then where the series disappears half way through.

Reply:
I see, you would need an `if len(series.Points) > 0` there instead of the `for`. I guess it's arguable whether that would be easier to read and/or more efficient so I guess I don't have a strong opinion.